src/Shared/NativeMethodsShared.cs

```diff
@@ -452,14 +468,52 @@ public SystemInformationData()
         #region Member data
 
         /// <summary>
-        /// Default buffer size to use when dealing with the Windows API.
+        /// Gets an enum for the max path restrictions of the current OS.
         /// </summary>
-        /// <remarks>
-        /// This member is intentionally not a constant because we want to allow
-        /// unit tests to change it.
-        /// </remarks>
-        internal static int MAX_PATH = 260;
+        internal static MaxPathLimits OSMaxPathLimit
+        {
+            get
+            {
+#if EXPERIMENTAL_LONGPATHS_ENABLED
+                if (osMaxPathLimit == MaxPathLimits.Unknown)
+                {
+                    SetOSMaxPathLimit();
+                }
+                return osMaxPathLimit;
+#else
+                return MaxPathLimits.LegacyWindows;
+#endif
+            }
+        }
+
+        /// <summary>
+        /// Cached value for OSMaxPathRestriction.
+        /// </summary>
+        private static MaxPathLimits osMaxPathLimit = MaxPathLimits.Unknown;
+
+        private static readonly object osMaxPathLimitLock = new object();
+
+        private static void SetOSMaxPathLimit()
+        {
+            lock (osMaxPathLimitLock)
+            {
+                if (osMaxPathLimit == MaxPathLimits.Unknown)
+                {
+                    osMaxPathLimit = IsMaxPathLimitLegacyWindows() ? MaxPathLimits.LegacyWindows : MaxPathLimits.None;
+                }
+            }
+        }
 
+        private static bool IsMaxPathLimitLegacyWindows()
+        {
+            if (!IsWindows)
+            {
+                return false;
+            }
+            RegistryKey fileSystemKey = Registry.LocalMachine.OpenSubKey(WINDOWS_FILE_SYSTEM_REGISTRY_KEY);
+            object longPathsEnabledValue = fileSystemKey.GetValue(WINDOWS_LONG_PATHS_ENABLED_VALUE_NAME, 0);
+            return Convert.ToInt32(longPathsEnabledValue) != 1;
```

Comment:
This could throw. We could set to legacy or none in that case and continue? Or expose to the user? Seems like silently doing the best we can would be good though for something like this.